code/datums/components/gunpoint.dm

```diff
@@ -0,0 +1,145 @@
+#define GUNPOINT_SHOOTER_STRAY_RANGE 3
+#define GUNPOINT_DELAY_STAGE_2 20
+#define GUNPOINT_DELAY_STAGE_3 80 // cumulative with past stages, so 100 deciseconds
+#define GUNPOINT_MULT_STAGE_1 1
+#define GUNPOINT_MULT_STAGE_2 2
+#define GUNPOINT_MULT_STAGE_3 2.5
+
+
+/datum/component/gunpoint
+	dupe_mode = COMPONENT_DUPE_ALLOWED
+
+	var/mob/living/target
+	var/obj/item/gun/weapon
+
+	var/stage = 1
+	var/datum/status_effect/status_hold_up
+	var/datum/status_effect/status_held_up
+	var/damage_mult = GUNPOINT_MULT_STAGE_1
+
+// *extremely bad russian accent* no!
+/datum/component/gunpoint/Initialize(mob/living/targ, obj/item/gun/wep)
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	var/mob/living/shooter = parent
+	target = targ
+	weapon = wep
+	RegisterSignal(targ, list(COMSIG_MOB_ATTACK_HAND, COMSIG_MOB_ITEM_ATTACK, COMSIG_MOB_ATTACK_RANGED, COMSIG_MOVABLE_MOVED), .proc/trigger_reaction)
```

Comment:
Double check that the MOB_ATTACK_RANGED signal doesn't trigger if the person unarmed clicks on a tile they can't reach. From what I can see the signal doesn't send if they fired a gun, but does if they attempted to unarmed click outside their melee range, which only works as intended here if they tried initiating (but apparently not throwing things with) telekinesis. Might also want to make throwing trigger the shot if throwing stuff is considered lethal enough to matter.

Reply:
blorble, I realized the issue with MOB_ATTACK_RANGED and thought I pushed removing it, but yeah I will probably follow your approach with adding a signal onto guns firing

also I'm gonna leave throwing unchecked since there can be legit reasons for it ("throw me your ID card!"), and any damage is a 50/50 chance to cause a flinch anyway so it works out as a tactical choice

Also re: cancelling if the gunman fires a gun, I'd rather not do that since it'd severely limit the hostage taking potential if other people could rush in knowing the gunman loses their edge if they fire. If it's problematic, I can change it afterwards